Staked Celo Devchain

soloseng opened the pull request:
### Description

Added a script to generate a new devchain tarball containing the deployed stacked-celo contracts.

See `README` on how to use.

### Other changes

In order to add new deployment to devchain tarball, I needed the current celo-devchain package to allow persistant storage of chain data. I made the necessary changes and submitted a [PR](https://github.com/terminal-fi/celo-devchain/issues/2).

In the mean time we are using a github [link](https://github.com/soloseng/celo-devchain/tree/with-db-option) to access the updated celo-devchain package.

### Tested

Added  test script to ensure proper deployment of staked CELO contracts to devchain.

### Related issues

- Fixes #5 
- Fixes #16

montera82 commented on package.json:
```diff
@@ -14,13 +13,13 @@
     "@nomiclabs/hardhat-web3": "^2.0.0",
     "@openzeppelin/contracts": "^4.4.2",
     "@openzeppelin/contracts-upgradeable": "^4.5.2",
+    "@terminal-fi/celo-devchain": "git+https://github.com/soloseng/celo-devchain#with-db-option",
```
nit: very innovative way to move forward despite delays, and do you think it'll be cool to fork this under the celo-org github organisation?

soloseng replied:
We could do that if we want to maintain our version of the `devchain` npm package. I guess we should evaluate that depending on response time of the original packages maintainers.

montera82 commented on scripts/test/devchain.test.ts:
```diff
@@ -0,0 +1,75 @@
+  let account: any;
+  let manager: any;
+  let stCELO: any;
+  let rstCELO: any;
```
any reason why we have these as any, or no typechain for those 🤔

soloseng replied:
Since this test is checking the proper deployment of staked CELO contracts on the devchain, for use by anyone in a new project. It assumes the staked CELO contract will not be compiled locally and, therefore, no typings generated before running the test.

m-chrzan replied:
But we're running this in this repo, right, so we do have the typings available?

montera82 commented on README.md:
```diff
@@ -102,6 +102,107 @@ Withdrawal flow:
+# Start with a fresh checkout to avoid build complications.
+> git clone https://github.com/celo-org/celo-monorepo.git
+> git fetch --all --tags
+> git checkout soloseng/add-devchain-option
```
so this is technically checking out a new branch right ..i.e checkout -b name-of-new-branch

soloseng replied:
Yes, it's currently checking out a branch that gives us the option to compress the chaindata. (this is temp until branch is merged with `master`.

@m-chrzan thinking more about it, I'm wondering if it might be a bit overkill to clone the `CELO-MONOREPO` only to compress some files vs a script in `STAKED-CELO` repo to handle the compression. Thoughts?

m-chrzan replied:
What do you think about having a simple wrapper script here around the monorepo devchain script that takes a `--monorepo` flag (which assumes the monorepo is already built and checked out at the correct branch)?

soloseng replied:
Added a script that assumes monorepo is available locally. Details on how to use in the README.

m-chrzan commented on README.md:
```diff
@@ -102,6 +102,108 @@ Withdrawal flow:
+> yarn build --ignore docs
+
+> cd packages/protocol
+> yarn devchain compress-chain <path_to_staked_CELO_chainData_dir> .tmp/stakedCeloDevchain.tar.gz
```
Why are we directing them to save in a `.tmp` directory?

m-chrzan commented on scripts/mineBlocks.ts:
```diff
@@ -0,0 +1,19 @@
+  devchainProvider = new hre.ethers.providers.JsonRpcProvider("http://localhost:7545");
+}
+
+async function mineNBlocks(n: number) {
```
Note there is already a `mineBlocks` function in `test-ts/utils.ts`. Could it be reused?

m-chrzan commented on scripts/test/devchain.test.ts:
```diff
@@ -0,0 +1,75 @@
+    expect(ownerList[2]).to.eq(multisigOwner2.address);
+  });
+
+  it("multisig first be owned by 3 addresses", async () => {
```
This looks like an incomplete copy of the previous test?

m-chrzan commented on scripts/test/devchain.test.ts:
```diff
@@ -0,0 +1,75 @@
+    expect(currentManager).to.eq(manager.address);
+  });
+
+  it("Account should have manager contract address as manager", async () => {
```
```suggestion
  it("Account should have Manager contract address as manager", async () => {
```

m-chrzan commented on scripts/test/devchain.test.ts:
```diff
@@ -0,0 +1,75 @@
+    expect(ownerList[1]).to.eq(multisigOwner1.address);
+  });
+
+  it("Manager should be owned by multisig", async () => {
```
```suggestion
  it("Manager should be owned by MultiSig", async () => {
```

m-chrzan commented on scripts/test/devchain.test.ts:
```diff
@@ -0,0 +1,75 @@
+    rstCELO = await hre.ethers.getContract("RebasedStakedCelo");
+  });
+
+  it("multisig should be owned by 3 addresses", async () => {
```
```suggestion
  it("MultiSig should be owned by 3 addresses", async () => {
```

m-chrzan commented on scripts/test/devchain.test.ts:
```diff
@@ -0,0 +1,75 @@
+  let stCELO: any;
+  let rstCELO: any;
+
+  let multisigOwner0: SignerWithAddress;
```
`multiSig` here and below

m-chrzan commented on scripts/test/devchain.test.ts:
```diff
@@ -0,0 +1,75 @@
+    expect(currentManager).to.eq(manager.address);
+  });
+
+  it("StakedCelo should have manager contract address as manager", async () => {
```
```suggestion
  it("StakedCelo should have Manager contract address as manager", async () => {
```

m-chrzan commented on scripts/tarchain.ts:
```diff
@@ -0,0 +1,74 @@
+      exitOnError(
+        runCmd(
+          args.datadir!,
+          args.filename != ("" || undefined) ? args.filename : "stakedCeloDevchain.tar.gz",
```
Doesn't `("" || undefined)` just always evaluate to `undefined`?

soloseng replied:
It never actually evaluates to `undefined`, as `yargs` parses empty `filename` argument as an empty string, but since its type is `string | undefined` I added an undefined check to make it null safe.

m-chrzan commented on scripts/tarchain.ts:
```diff
@@ -0,0 +1,74 @@
+
+  fs.copySync(deploymentsDir, chainDataDir + deploymentsDir);
+  const cmdArgs = ["devchain", "compress-chain", datadir, chainDataDir + filename];
+  const ProtocolRoot = path.normalize(path.join(__dirname, `../${monorepo}/packages/protocol`));
```
Why is there a `..` before the `monorepo` value? I would assume `monorepo` is meant to point directly at the monorepo already.

soloseng replied:
Used `process.cwd()` instead of `__dirname`. This way `protocolRoot` is based on the `cwd` instead of the location of `tarchain.ts`.

This change should prevent breakage if the `tarchain.ts` get nested further.

m-chrzan replied:
I guess I don't understand why you're `join`ing anything to the already fully specified monorepo path. E.g. if `cwd` is `/home/martin/staked-celo` and I provide the monorepo path as an absolute path, e.g. `/home/martin/celo-monorepo`, the current code would lead to `/home/martin/staked-celo/home/martin/celo-monorepo`, right? Which doesn't exist and isn't what we want.

soloseng replied:
I joined them since I was using relative path. 

I can change it to using absolute path instead.

soloseng replied:
changed to absolute path

m-chrzan replied:
I mean, this should now work with either absolute or relative paths. If I do `--monorepo ../<rest of path>` or `--monorepo /<absolute path>`, this should work in either case. Before, it was essentially required you'd provide a relative path, relative to the directory higher up, rather than relative to the current directory. Right?

m-chrzan commented on scripts/tarchain.ts:
```diff
@@ -0,0 +1,74 @@
+  options?: SpawnOptions & { silent?: boolean }
+) {
+  return new Promise<number>(async (resolve, reject) => {
+    const { silent, ...spawnOptions } = options || { silent: false };
```
Looks like we're always setting `options` to just a `{ cwd: ... }` object (last line of `runCmd`), is there a code path where `options` is undefined and we end up taking the right side of this `||`? Should all this stuff around `silent` be removed, or should an option for selecting whether we want to run silently or not be added?

soloseng replied:
Removed the `silent` option, as there is already sufficient error logging without it.

m-chrzan commented on scripts/test/devchain.test.ts:
```diff
@@ -0,0 +1,75 @@
+  });
+
+  it("Manager should be owned by MultiSig", async () => {
+    const currentManager = await account.manager();
+    expect(currentManager).to.eq(manager.address);
```
This checks same thing as below test, not that MultiSig is owner.

m-chrzan replied:
Once this test is correct, similar could be done for Account, StakedCelo, and each Proxy.

soloseng commented:
Made two changes

1. Added a `devchain` network with `port 7545` to the `hardhat.config.ts` and kept `local` with `port 8545` for use with any other local node that we may be running
2. Changed the `tarchain` command to use `--datadir`  & ` --monorepo` flags instead of positional arguments

m-chrzan commented on .gitignore:
```diff
@@ -8,9 +8,11 @@ node_modules/
 cache
-deployments/localhost
-deployments/hardhat
+/deployments
```
This will match the absolute path `/deployments`, no? Which is probably not what we want.

soloseng replied:
Will we be committing our testnet or mainnet deployment files?

m-chrzan replied:
I think that's the convention with hardhat projects, no? You want to keep around the `deployments/network` directory as it keeps states that's necessary for future deployments to run, right?

soloseng replied:
Makes sense, specially that we are using upgradable contracts. I made the changes to explicitly ignored the local networks only.
